Plugins/Events/Events/ItemEvents.cpp

```diff
@@ -149,4 +154,80 @@ void ItemEvents::RemoveItemHook(Services::Hooks::CallType type, CItemRepository*
     Events::SignalEvent(before ? "NWNX_ON_ITEM_INVENTORY_REMOVE_ITEM_BEFORE" : "NWNX_ON_ITEM_INVENTORY_REMOVE_ITEM_AFTER", thisPtr->m_oidParent);
 }
 
+uint32_t ItemEvents::FindItemWithBaseItemIdHook(CItemRepository* thisPtr, uint32_t baseItem, int32_t nTh)
+{
+    auto *pItemHolder = Utils::AsNWSCreature(Globals::AppManager()->m_pServerExoApp->GetGameObject(thisPtr->m_oidParent));
+
+    if(!pItemHolder)
+    {
+        // For our purposes we only want this to be used on creature ItemRepositories
+        return m_FindItemWithBaseItemIdHook->CallOriginal<int32_t>(thisPtr, baseItem, nTh);
+    }
+
+    // We're most assuredly in an infinite loop here caused by the scripting end
+    // unless the player has > 255 stacks of ammo they can't equip
+    // Suppose I could set this to a higher value but I think this is safe
+    if(nTh > 255)
+    {
+        return OBJECT_INVALID;
+    }
+
+    auto ItemSanityCheck = [&](uint32_t objectId) -> uint32_t {
+        if (static_cast<Types::ObjectID>(objectId) == Constants::OBJECT_INVALID)
+            return objectId;
+
+        auto *pItem = Utils::AsNWSItem(Globals::AppManager()->m_pServerExoApp->GetGameObject(objectId));
+        if (!pItem)
+        {
+            LOG_WARNING("Item does not exist, falling back to original call.");
+            objectId = m_FindItemWithBaseItemIdHook->CallOriginal<uint32_t>(thisPtr, baseItem, nTh);
+        }
+        else if (pItem->m_nBaseItem != baseItem)
+        {
+            LOG_WARNING("Base Item ID of returned item does not match, falling back to original call.");
+            objectId = m_FindItemWithBaseItemIdHook->CallOriginal<uint32_t>(thisPtr, baseItem, nTh);
+        }
+        else if (pItem->m_oidPossessor != thisPtr->m_oidParent)
+        {
+            LOG_WARNING("Item does not belong to that creature, falling back to original call.");
+            objectId = m_FindItemWithBaseItemIdHook->CallOriginal<uint32_t>(thisPtr, baseItem, nTh);
+        }
+        // TODO: Check if player is allowed to equip the item
+        return objectId;
+    };
+
+    std::string sBeforeEventResult;
+    std::string sAfterEventResult;
+
+    uint32_t retVal;
+
+    Events::PushEventData("BASE_ITEM_ID", std::to_string(baseItem));
+    Events::PushEventData("BASE_ITEM_NTH", std::to_string(nTh));
+
+    if (Events::SignalEvent("NWNX_ON_ITEM_FIND_ITEM_WITH_BASE_ITEMID_BEFORE", thisPtr->m_oidParent, &sBeforeEventResult))
```

Comment:
This seems like an overly generic event on an overly generic function that just so happens to not be used much currently. I'm thinking just rename the event to `NWNX_ON_AMMO_RELOAD` or something, and add an early return to the function if the base item is not one of the ammo types.

That way, future changes to the game won't cause this to randomly fire in other cases.